cdap-app-fabric/src/main/java/co/cask/cdap/extension/AbstractExtensionLoader.java

```diff
@@ -0,0 +1,225 @@
+import com.google.common.collect.Iterables;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.ServiceLoader;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
+
+/**
+ * A class to load extensions for supported extension types from a configured extensions directory. It uses the Java
+ * {@link ServiceLoader} architecture to load extensions from the specified directory. It first tries to load
+ * extensions using a {@link URLClassLoader} consisting of all jars in the configured extensions directory. For unit
+ * tests, it loads extensions using the system classloader. If an extension is not found via the extensions classloader
+ * or the system classloader, it returns a default extension.
+ *
+ * The supported directory structure is:
+ * <pre>
+ *   [extensions-directory]/[extensions-module-directory-1]/
+ *   [extensions-directory]/[extensions-module-directory-1]/file-containing-extensions-1.jar
+ *   [extensions-directory]/[extensions-module-directory-1]/file-containing-dependencies-1.jar
+ *   [extensions-directory]/[extensions-module-directory-1]/file-containing-dependencies-2.jar
+ *   [extensions-directory]/[extensions-module-directory-2]/
+ *   [extensions-directory]/[extensions-module-directory-2]/file-containing-extensions-2.jar
+ *   [extensions-directory]/[extensions-module-directory-2]/file-containing-extensions-3.jar
+ *   [extensions-directory]/[extensions-module-directory-2]/file-containing-dependencies-3.jar
+ *   [extensions-directory]/[extensions-module-directory-2]/file-containing-dependencies-4.jar
+ *   ...
+ *   [extensions-directory]/[extensions-module-directory-n]/file-containing-extensions-n.jar
+ * </pre>
+ *
+ * Each extensions jar file above can contain multiple extensions.
+ *
+ *
+ *
+ * @param <EXTENSION_TYPE> the data type of the objects that a given extension can be used for. e.g. for a program
+ *                        runtime extension, the list of program types that the extension can be used for
+ * @param <EXTENSION> the data type of the extension
+ */
+public abstract class AbstractExtensionLoader<EXTENSION_TYPE, EXTENSION> {
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractExtensionLoader.class);
+
+  private final List<String> extDirs;
+  private final Class<EXTENSION> extensionClass;
+  // A ServiceLoader that loads extension implementation from the CDAP system classloader.
+  private final ServiceLoader<EXTENSION> systemExtensionLoader;
+  private final LoadingCache<EXTENSION_TYPE, AtomicReference<EXTENSION>> extensionsCache;
+
+  public AbstractExtensionLoader(String extDirs, Class<EXTENSION> extensionClass) {
```

Comment:
Instead of taking a `Class` from the constructor, assuming the sub-class has the type set, you can actually get it using `TypeToken` like this:

```java
    Type type = TypeToken.of(getClass()).getSupertype(AbstractExtensionLoader.class).getType();
    if (!(type instanceof ParameterizedType)) {
      // Can't happen
      throw new IllegalStateException();
    }
    Type extensionType = ((ParameterizedType) type).getActualTypeArguments()[1];
    if (!(extensionType instanceof Class)) {
      // Can't happen
      throw new IllegalStateException();
    }
    this.extensionClass = (Class<EXTENSION>) extensionType;
```